spring-vault-core/src/main/java/org/springframework/vault/authentication/AppRoleAuthenticationOptions.java

```diff
@@ -148,9 +207,16 @@ public AppRoleAuthenticationOptionsBuilder secretId(String secretId) {
 		public AppRoleAuthenticationOptions build() {
 
 			Assert.hasText(path, "Path must not be empty");
-			Assert.notNull(roleId, "RoleId must not be null");
 
-			return new AppRoleAuthenticationOptions(path, roleId, secretId);
+			//Need to have either RoleID or (AppRole and RoleToken)
+			if(StringUtils.isEmpty(appRole) && StringUtils.isEmpty(roleToken)){
```

Comment:
How about splitting these `if`s and decouple the cases? `if`...`else` is quite hard to read. We should also verify the various combinations (roleId/secretId are both present -> token/role name not required, roleId or secretId missing -> token/role must be present).

Reply:
I'll try to make it easier to read.
Checking the actual implementation, secretId can be ignored if roleId is set.
The combinations I am thinking about are:
- appRole and initialToken are empty -> we need a roleId to authenticate
- roleId is empty -> I need a appRole and a token to get at least the roleId

Reply:
You're right, secretId is optional – When it's empty and a token is present then it's the pull case to fetch the secretId before authentication.

Reply:
I have changed the initial check to reflect this logic